azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+                    module_id=auth_provider.module_id,
+                    hostname=auth_provider.hostname,
+                    gateway_hostname=gateway_hostname,
+                    ca_cert=auth_provider.ca_cert,
+                ),
+                pipeline_ops.Iot_SetSasToken(sas_token=auth_provider.get_current_sas_token()),
+                callback=op.callback,
+            )
+        else:
+            self.continue_op(op)
+
+
+# TODO: make sure everything that is private is named with underscore
+
+
+class Iot_EnsureConnection(PipelineStage):
+    """
+    This stage is responsible for ensuring that the transport is connected when
+    it needs to be connected, and it's responsible for queueing operations
+    while we're waiting for the connect operation to complete.
+
+    Operations Handled:
+    * Iot_Connect
+    * Iot_Disconnect
+    * all operations with needs_connection set to True (connects if necessary)
+    * all operations (queues while waiting for connection)
+
+    Operations Produced:
+    * Iot_Connect
+
+    Note: this stage will likely be replaced by a more full-featured stage to handle
+    other "block while we're setting something up" operations, such as subscribing to
+    twin responses.  That is another example where we want to ensure some state and block
+    requests until that state is achieved.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.connected = False
+        self.queue = deque()
+        self.blocked = False
+
+    def _run_op(self, op):
+        logger.info("{}({}) handling".format(self.name, op.name))
+
+        # If this stage is currently blocked (because we're waiting for a connection
+        # to complete, we queue up all operations until after the connect completes.
+        if self.blocked:
+            logger.info("pipeline is blocked waiting for connect.  queueing.")
+            self.queue.append(op)
+
+        # If we get a request to connect, we either complete immediately (if we're already
+        # connected) or we do the connect operation, which is pulled out into a helper
+        # function because starting the connection also means blocking this stage until
+        # the connect is complete.
+        elif op.name is pipeline_ops.names.Iot_Connect:
+            if self.connected:
+                logger.info("transport is already conencted.  completing early.")
+                self.complete_op(op)
+            else:
+                self._do_connect(op)
+
+        # If we get a request to disconnect, we either complete the request immediately
+        # (if we're already disconencted) or we pass the disconnect request down.
+        elif op.name is pipeline_ops.names.Iot_Disconnect:
+            if not self.connected:
+                logger.info("transport is already disconencted.  completing early.")
+                self.complete_op(op)
+            else:
+                self.continue_op(op)
```

Comment:
Admittedly though, it is a bit semantically weird to be handling disconnects in the EnsureConnection stage. Of course, you have to handle disconnects in ensure connection, but I'd like it to clearly trigger a reconnect here. I'm guessing that's a side effect of disconnect, but it isn't clear from the scope of this class, and thus the semantics are odd to me.

Reply:
I think reconnecting is in a different not-yet-implemented retry stage.  

This stage is going to be heavily refactored into the stage we discussed at the whiteboard that ensures connection and also ensures subscriptions.   Ops are going to have an array of "demands" or 'requirements" that they need before they can run, and this stage will enforce those demands.  "connection" is one requirement for operations that require a connection, other requirements will all be subscriptions, at least to start., so an op could say that it requires a connection and a subscription to the twin topics before it can run.

---
In reply to: [279675524](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279675524) [](ancestors = 279675524)